Error using discovery strategy in hazelcast client

Version: 3.8.4
Cluster Size: 2
Clients: 1
Java 8

Hi,

I've come across an issue(s) testing failover over the client. I have a very simple 2 member cluster running on docker/mesosphere. The cluster itself uses a discovery strategy to discover it's cluster peers (via marathon). This works very well. 

I also use a discovery strategy in the hazlecast client by setting a discovery strategy that just calls a restful service returning the address/port of the cluster members.

My client configuration looks like:

```java
ClientConfig config = new ClientConfig();		
DiscoveryConfig discoveryConfig = new DiscoveryConfig();
		discoveryConfig.addDiscoveryStrategyConfig(
				new DiscoveryStrategyConfig(new ClusterDiscoveryStrategyFactory(URI.create(hazelcastClusterDiscoveryUri))));
		ClientNetworkConfig networkConfig = new ClientNetworkConfig();
		networkConfig.setDiscoveryConfig(discoveryConfig);
		config.setNetworkConfig(networkConfig);		
		config.setProperty("hazelcast.discovery.enabled", "true");
        config.setProperty("hazelcast.logging.type", "slf4j");
		config.setSerializationConfig(new SerializationConfig()
				.setGlobalSerializerConfig(new GlobalSerializerConfig()
						.setImplementation(new HazelcastKryoSerializer<>(new KryoSerializer(false)))));
        return HazelcastClient.newHazelcastClient(config);
```

My first issue is that even though I set the discovery strategy and don't explicitly set the addresses for the client to connect to, Hazelcast adds a DefaultAddressProvider that adds localhost. Is this by design or a configuration issue on my side? I am providing a discovery strategy to return the cluster members to connect to. I don't want hazelcast to attempt to bind to 5701/2/3 on localhost.

My second issue is that, in my tests, when I stop the cluster member that the client is connected to, I expect the client to seamlessly failover to the remaining member in the cluster. What I am seeing is that, on startup, the client is establishing a connection with both cluster members. The first one is the owner member

```log
2017-08-16 16:43:53,676 INFO [hz.client_0.internal-2] c.h.c.c.ClientConnectionManager [?:?] hz.client_0 [dev] [3.8.4] Setting ClientConnection{alive=true, connectionId=1, socketChannel=DefaultSocketChannelWrapper{socketChannel=java.nio.channels.SocketChannel[connected local=/10.64.0.29:64800 remote=/89.21.121.14:31765]}, remoteEndpoint=[lx-mesos-d02.unix.dnsbego.de]:31765, lastReadTime=2017-08-16 16:43:53.648, lastWriteTime=2017-08-16 16:43:53.613, closedTime=never, lastHeartbeatRequested=never, lastHeartbeatReceived=never, connected server version=3.8.3} as owner  with principal ClientPrincipal{uuid='56c4d015-4ad3-4bf4-82ef-57a23c5b2b64', ownerUuid='69713e18-22d1-4bca-bf83-019d1411a66c'}
2017-08-16 16:43:53,677 TRACE [hz.client_0.internal-2] c.h.c.c.ClientConnectionManager [?:?] hz.client_0 [dev] [3.8.4] Authentication succeeded for ClientConnection{alive=true, connectionId=1, socketChannel=DefaultSocketChannelWrapper{socketChannel=java.nio.channels.SocketChannel[connected local=/10.64.0.29:64800 remote=/89.21.121.14:31765]}, remoteEndpoint=[lx-mesos-d02.unix.dnsbego.de]:31765, lastReadTime=2017-08-16 16:43:53.648, lastWriteTime=2017-08-16 16:43:53.613, closedTime=never, lastHeartbeatRequested=never, lastHeartbeatReceived=never, connected server version=3.8.3} and there was no old connection to this end-point
2017-08-16 16:43:53,677 INFO [hz.client_0.internal-2] c.h.c.c.ClientConnectionManager [?:?] hz.client_0 [dev] [3.8.4] Authenticated with server [lx-mesos-d02.unix.dnsbego.de]:31765, server version:3.8.3 Local address: /10.64.0.29:64800
2017-08-16 16:43:53,691 INFO [hz.client_0.event-2] c.h.c.s.i.ClientMembershipListener [?:?] hz.client_0 [dev] [3.8.4] 

Members [2] {
	Member [lx-mesos-d13.unix.dnsbego.de]:31744 - bbe23fd9-a5ef-424b-8fdb-48b7446da79a
	Member [lx-mesos-d02.unix.dnsbego.de]:31765 - 69713e18-22d1-4bca-bf83-019d1411a66c
}

2017-08-16 16:43:53,692 INFO [main] c.h.c.LifecycleService [?:?] hz.client_0 [dev] [3.8.4] HazelcastClient 3.8.4 (20170809 - 297a77e) is CLIENT_CONNECTED
2017-08-16 16:43:53,810 TRACE [hz.client_0.internal-2] c.h.c.c.ClientConnectionManager [?:?] hz.client_0 [dev] [3.8.4] Authentication succeeded for ClientConnection{alive=true, connectionId=2, socketChannel=DefaultSocketChannelWrapper{socketChannel=java.nio.channels.SocketChannel[connected local=/10.64.0.29:64801 remote=lx-mesos-d13.unix.dnsbego.de/89.21.122.52:31744]}, remoteEndpoint=[lx-mesos-d13.unix.dnsbego.de]:31744, lastReadTime=2017-08-16 16:43:53.809, lastWriteTime=2017-08-16 16:43:53.780, closedTime=never, lastHeartbeatRequested=never, lastHeartbeatReceived=never, connected server version=3.8.3} and there was no old connection to this end-point
2017-08-16 16:43:53,811 INFO [hz.client_0.internal-2] c.h.c.c.ClientConnectionManager [?:?] hz.client_0 [dev] [3.8.4] Authenticated with server [lx-mesos-d13.unix.dnsbego.de]:31744, server version:3.8.3 Local address: /10.64.0.29:64801
```
From the logs above, it has connected to and authenticated against both members.
Now, I kill the cluster member that the client is connected to:
```log
2017-08-16 16:43:55,199 WARN [hz.hz.client_0.IO.thread-in-0] c.h.c.c.n.ClientConnection [?:?] hz.client_0 [dev] [3.8.4] ClientConnection{alive=false, connectionId=1, socketChannel=DefaultSocketChannelWrapper{socketChannel=java.nio.channels.SocketChannel[connected local=/10.64.0.29:64800 remote=/89.21.121.14:31765]}, remoteEndpoint=[lx-mesos-d02.unix.dnsbego.de]:31765, lastReadTime=2017-08-16 16:43:55.196, lastWriteTime=2017-08-16 16:43:55.187, closedTime=2017-08-16 16:43:55.197, lastHeartbeatRequested=never, lastHeartbeatReceived=never, connected server version=3.8.3} lost. Reason: java.io.EOFException[Remote socket closed!]
2017-08-16 16:43:55,200 INFO [hz.hz.client_0.IO.thread-in-0] c.h.c.c.ClientConnectionManager [?:?] hz.client_0 [dev] [3.8.4] Removed connection to endpoint: [lx-mesos-d02.unix.dnsbego.de]:31765, connection: ClientConnection{alive=false, connectionId=1, socketChannel=DefaultSocketChannelWrapper{socketChannel=java.nio.channels.SocketChannel[closed]}, remoteEndpoint=[lx-mesos-d02.unix.dnsbego.de]:31765, lastReadTime=2017-08-16 16:43:55.196, lastWriteTime=2017-08-16 16:43:55.199, closedTime=2017-08-16 16:43:55.197, lastHeartbeatRequested=never, lastHeartbeatReceived=never, connected server version=3.8.3}
2017-08-16 16:43:55,200 INFO [hz.hz.client_0.IO.thread-in-0] c.h.c.LifecycleService [?:?] hz.client_0 [dev] [3.8.4] HazelcastClient 3.8.4 (20170809 - 297a77e) is CLIENT_DISCONNECTED
```
After this I would expect the existing connection made on startup to be promoted in the client. However, it initiates a new connection (with a new connection ID) and thus breaks the assertion in the code as below:
```log
2017-08-16 16:43:56,277 INFO [hz.client_0.cluster-] c.h.c.s.i.ClusterListenerSupport [?:?] hz.client_0 [dev] [3.8.4] Trying to connect to [89.21.122.52]:31744 as owner member
2017-08-16 16:43:56,287 INFO [hz.client_0.internal-3] c.h.c.c.ClientConnectionManager [?:?] hz.client_0 [dev] [3.8.4] Setting ClientConnection{alive=true, connectionId=3, socketChannel=DefaultSocketChannelWrapper{socketChannel=java.nio.channels.SocketChannel[connected local=/10.64.0.29:64808 remote=/89.21.122.52:31744]}, remoteEndpoint=[lx-mesos-d13.unix.dnsbego.de]:31744, lastReadTime=2017-08-16 16:43:56.286, lastWriteTime=2017-08-16 16:43:56.281, closedTime=never, lastHeartbeatRequested=never, lastHeartbeatReceived=never, connected server version=3.8.3} as owner  with principal ClientPrincipal{uuid='56c4d015-4ad3-4bf4-82ef-57a23c5b2b64', ownerUuid='bbe23fd9-a5ef-424b-8fdb-48b7446da79a'}
2017-08-16 16:43:56,287 TRACE [hz.client_0.internal-3] c.h.c.c.ClientConnectionManager [?:?] hz.client_0 [dev] [3.8.4] Re-authentication succeeded for ClientConnection{alive=true, connectionId=3, socketChannel=DefaultSocketChannelWrapper{socketChannel=java.nio.channels.SocketChannel[connected local=/10.64.0.29:64808 remote=/89.21.122.52:31744]}, remoteEndpoint=[lx-mesos-d13.unix.dnsbego.de]:31744, lastReadTime=2017-08-16 16:43:56.286, lastWriteTime=2017-08-16 16:43:56.281, closedTime=never, lastHeartbeatRequested=never, lastHeartbeatReceived=never, connected server version=3.8.3}
2017-08-16 16:43:56,288 ERROR [hz.client_0.internal-3] c.h.c.s.ClientInvocationService [Slf4jFactory.java:68] hz.client_0 [dev] [3.8.4] Failed asynchronous execution of execution callback: com.hazelcast.client.spi.impl.ClientInvocationFuture$InternalDelegatingExecutionCallback@e612fa9for call ClientMessage{length=141, correlationId=14, messageType=2, partitionId=-1, isComplete=true, isRetryable=true, isEvent=false, writeOffset=0}
java.lang.AssertionError: null
	at com.hazelcast.client.connection.nio.ClientConnectionManagerImpl.onAuthenticated(ClientConnectionManagerImpl.java:669) ~[hazelcast-client-3.8.4.jar:3.8.4]
	at com.hazelcast.client.connection.nio.ClientConnectionManagerImpl.access$900(ClientConnectionManagerImpl.java:91) ~[hazelcast-client-3.8.4.jar:3.8.4]
	at com.hazelcast.client.connection.nio.ClientConnectionManagerImpl$2.onResponse(ClientConnectionManagerImpl.java:578) ~[hazelcast-client-3.8.4.jar:3.8.4]
	at com.hazelcast.client.connection.nio.ClientConnectionManagerImpl$2.onResponse(ClientConnectionManagerImpl.java:550) ~[hazelcast-client-3.8.4.jar:3.8.4]
	at com.hazelcast.client.spi.impl.ClientInvocationFuture$InternalDelegatingExecutionCallback.onResponse(ClientInvocationFuture.java:129) ~[hazelcast-client-3.8.4.jar:3.8.4]
	at com.hazelcast.client.spi.impl.ClientInvocationFuture$InternalDelegatingExecutionCallback.onResponse(ClientInvocationFuture.java:117) ~[hazelcast-client-3.8.4.jar:3.8.4]
	at com.hazelcast.client.spi.impl.ClientInvocationFuture$InternalDelegatingExecutionCallback.onResponse(ClientInvocationFuture.java:129) ~[hazelcast-client-3.8.4.jar:3.8.4]
	at com.hazelcast.client.spi.impl.ClientInvocationFuture$InternalDelegatingExecutionCallback.onResponse(ClientInvocationFuture.java:117) ~[hazelcast-client-3.8.4.jar:3.8.4]
	at com.hazelcast.spi.impl.AbstractInvocationFuture$1.run(AbstractInvocationFuture.java:248) ~[hazelcast-3.8.4.jar:3.8.4]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_05]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_05]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [na:1.8.0_05]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [na:1.8.0_05]
	at com.hazelcast.util.executor.LoggingScheduledExecutor$LoggingDelegatingFuture.run(LoggingScheduledExecutor.java:138) [hazelcast-3.8.4.jar:3.8.4]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_05]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_05]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_05]
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64) [hazelcast-3.8.4.jar:3.8.4]
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80) [hazelcast-3.8.4.jar:3.8.4]
```
As I read the implementation, the issue seems to be, if the owning member fails, it is establishing a new connection to an address where there is already an active connection

Any hints on how to resolve would be much appreciated

Thanks

--- a/hazelcast-client/src/main/java/com/hazelcast/client/connection/AddressProvider.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/connection/AddressProvider.java
@@ -16,7 +16,8 @@
 
 package com.hazelcast.client.connection;
 
-import java.net.InetSocketAddress;
+import com.hazelcast.nio.Address;
+
 import java.util.Collection;
 
 /**
@@ -25,8 +26,8 @@ import java.util.Collection;
 public interface AddressProvider {
 
     /**
-     * @return Collection of InetSocketAddress
+     * @return The possible member addresses to connect to.
      */
-    Collection<InetSocketAddress> loadAddresses();
+    Collection<Address> loadAddresses();
 
 }

--- a/hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnectionManagerImpl.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnectionManagerImpl.java
@@ -813,13 +813,13 @@ public class ClientConnectionManagerImpl implements ClientConnectionManager, Con
             attempt++;
             long nextTry = Clock.currentTimeMillis() + connectionAttemptPeriod;
 
-            Collection<Address> socketAddresses = getSocketAddresses();
-            for (Address inetSocketAddress : socketAddresses) {
+            Collection<Address> addresses = getPossibleMemberAddresses();
+            for (Address address : addresses) {
                 if (!client.getLifecycleService().isRunning()) {
                     throw new IllegalStateException("Giving up on retrying to connect to cluster since client is shutdown.");
                 }
-                triedAddresses.add(inetSocketAddress);
-                if (connectAsOwner(inetSocketAddress) != null) {
+                triedAddresses.add(address);
+                if (connectAsOwner(address) != null) {
                     return;
                 }
             }
@@ -877,7 +877,7 @@ public class ClientConnectionManagerImpl implements ClientConnectionManager, Con
 
     }
 
-    private Collection<Address> getSocketAddresses() {
+    private Collection<Address> getPossibleMemberAddresses() {
         final List<Address> addresses = new LinkedList<Address>();
 
         Collection<Member> memberList = client.getClientClusterService().getMemberList();

--- a/hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnectionManagerImpl.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnectionManagerImpl.java
@@ -420,10 +420,9 @@ public class ClientConnectionManagerImpl implements ClientConnectionManager, Con
         return connection;
     }
 
-    private Connection connectAsOwner(InetSocketAddress ownerInetSocketAddress) {
+    private Connection connectAsOwner(Address address) {
         Connection connection = null;
         try {
-            Address address = new Address(ownerInetSocketAddress);
             logger.info("Trying to connect to " + address + " as owner member");
             connection = getOrConnect(address, true);
             client.onClusterConnect(connection);
@@ -432,9 +431,9 @@ public class ClientConnectionManagerImpl implements ClientConnectionManager, Con
             fireConnectionEvent(LifecycleEvent.LifecycleState.CLIENT_CONNECTED);
             connectionStrategy.onConnectToCluster();
         } catch (Exception e) {
-            logger.warning("Exception during initial connection to " + ownerInetSocketAddress + ", exception " + e);
+            logger.warning("Exception during initial connection to " + address + ", exception " + e);
             if (null != connection) {
-                connection.close("Could not connect to " + ownerInetSocketAddress + " as owner", e);
+                connection.close("Could not connect to " + address + " as owner", e);
             }
             return null;
         }
@@ -808,14 +807,14 @@ public class ClientConnectionManagerImpl implements ClientConnectionManager, Con
 
     private void connectToClusterInternal() {
         int attempt = 0;
-        Set<InetSocketAddress> triedAddresses = new HashSet<InetSocketAddress>();
+        Set<Address> triedAddresses = new HashSet<Address>();
 
         while (attempt < connectionAttemptLimit) {
             attempt++;
             long nextTry = Clock.currentTimeMillis() + connectionAttemptPeriod;
 
-            Collection<InetSocketAddress> socketAddresses = getSocketAddresses();
-            for (InetSocketAddress inetSocketAddress : socketAddresses) {
+            Collection<Address> socketAddresses = getSocketAddresses();
+            for (Address inetSocketAddress : socketAddresses) {
                 if (!client.getLifecycleService().isRunning()) {
                     throw new IllegalStateException("Giving up on retrying to connect to cluster since client is shutdown.");
                 }
@@ -878,23 +877,23 @@ public class ClientConnectionManagerImpl implements ClientConnectionManager, Con
 
     }
 
-    private Collection<InetSocketAddress> getSocketAddresses() {
-        final List<InetSocketAddress> socketAddresses = new LinkedList<InetSocketAddress>();
+    private Collection<Address> getSocketAddresses() {
+        final List<Address> addresses = new LinkedList<Address>();
 
         Collection<Member> memberList = client.getClientClusterService().getMemberList();
         for (Member member : memberList) {
-            socketAddresses.add(member.getSocketAddress());
+            addresses.add(member.getAddress());
         }
 
         for (AddressProvider addressProvider : addressProviders) {
-            socketAddresses.addAll(addressProvider.loadAddresses());
+            addresses.addAll(addressProvider.loadAddresses());
         }
 
         if (shuffleMemberList) {
-            Collections.shuffle(socketAddresses);
+            Collections.shuffle(addresses);
         }
 
-        return socketAddresses;
+        return addresses;
     }
 
     private ExecutorService createSingleThreadExecutorService(HazelcastClientInstanceImpl client) {

--- a/hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/AwsAddressProvider.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/AwsAddressProvider.java
@@ -22,8 +22,8 @@ import com.hazelcast.client.connection.AddressProvider;
 import com.hazelcast.client.util.AddressHelper;
 import com.hazelcast.logging.ILogger;
 import com.hazelcast.logging.LoggingService;
+import com.hazelcast.nio.Address;
 
-import java.net.InetSocketAddress;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
@@ -44,10 +44,10 @@ public class AwsAddressProvider implements AddressProvider {
     }
 
     @Override
-    public Collection<InetSocketAddress> loadAddresses() {
+    public Collection<Address> loadAddresses() {
         updateLookupTable();
         final Map<String, String> lookupTable = getLookupTable();
-        final Collection<InetSocketAddress> addresses = new ArrayList<InetSocketAddress>(lookupTable.size());
+        final Collection<Address> addresses = new ArrayList<Address>(lookupTable.size());
 
         for (String privateAddress : lookupTable.keySet()) {
             addresses.addAll(AddressHelper.getSocketAddresses(privateAddress));

--- a/hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/DefaultAddressProvider.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/DefaultAddressProvider.java
@@ -19,8 +19,8 @@ package com.hazelcast.client.spi.impl;
 import com.hazelcast.client.config.ClientNetworkConfig;
 import com.hazelcast.client.connection.AddressProvider;
 import com.hazelcast.client.util.AddressHelper;
+import com.hazelcast.nio.Address;
 
-import java.net.InetSocketAddress;
 import java.util.Collection;
 import java.util.LinkedList;
 import java.util.List;
@@ -41,16 +41,16 @@ public class DefaultAddressProvider implements AddressProvider {
     }
 
     @Override
-    public Collection<InetSocketAddress> loadAddresses() {
+    public Collection<Address> loadAddresses() {
         final List<String> addresses = networkConfig.getAddresses();
         if (addresses.isEmpty() && noOtherAddressProviderExist) {
-            addresses.add("localhost");
+            addresses.add("127.0.0.1");
         }
-        final List<InetSocketAddress> socketAddresses = new LinkedList<InetSocketAddress>();
+        final List<Address> possibleAddresses = new LinkedList<Address>();
 
         for (String address : addresses) {
-            socketAddresses.addAll(AddressHelper.getSocketAddresses(address));
+            possibleAddresses.addAll(AddressHelper.getSocketAddresses(address));
         }
-        return socketAddresses;
+        return possibleAddresses;
     }
 }

--- a/hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/discovery/DiscoveryAddressProvider.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/discovery/DiscoveryAddressProvider.java
@@ -23,8 +23,6 @@ import com.hazelcast.nio.Address;
 import com.hazelcast.spi.discovery.DiscoveryNode;
 import com.hazelcast.spi.discovery.integration.DiscoveryService;
 
-import java.net.InetSocketAddress;
-import java.net.UnknownHostException;
 import java.util.ArrayList;
 import java.util.Collection;
 
@@ -42,18 +40,13 @@ public class DiscoveryAddressProvider
     }
 
     @Override
-    public Collection<InetSocketAddress> loadAddresses() {
+    public Collection<Address> loadAddresses() {
         Iterable<DiscoveryNode> discoveredNodes = checkNotNull(discoveryService.discoverNodes(),
                 "Discovered nodes cannot be null!");
 
-        Collection<InetSocketAddress> possibleMembers = new ArrayList<InetSocketAddress>();
+        Collection<Address> possibleMembers = new ArrayList<Address>();
         for (DiscoveryNode discoveryNode : discoveredNodes) {
-            Address discoveredAddress = discoveryNode.getPrivateAddress();
-            try {
-                possibleMembers.add(discoveredAddress.getInetSocketAddress());
-            } catch (UnknownHostException e) {
-                logger.warning("Unresolvable host exception", e);
-            }
+            possibleMembers.add(discoveryNode.getPrivateAddress());
         }
         return possibleMembers;
     }

--- a/hazelcast-client/src/main/java/com/hazelcast/client/util/AddressHelper.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/util/AddressHelper.java
@@ -21,9 +21,14 @@ import com.hazelcast.nio.Address;
 import com.hazelcast.util.AddressUtil;
 import com.hazelcast.util.AddressUtil.AddressHolder;
 
+import java.net.Inet4Address;
+import java.net.Inet6Address;
+import java.net.InetAddress;
 import java.net.UnknownHostException;
 import java.util.Collection;
 import java.util.LinkedList;
+
+import static com.hazelcast.util.AddressUtil.getPossibleInetAddressesFor;
 
 /**
  * This is a client side utility class for working with addresses and cluster connections
@@ -52,16 +57,37 @@ public final class AddressHelper {
 
     public static Collection<Address> getPossibleSocketAddresses(int port, String scopedAddress,
                                                                            int portTryCount) {
+        InetAddress inetAddress = null;
+        try {
+            inetAddress = InetAddress.getByName(scopedAddress);
+        } catch (UnknownHostException ignored) {
+            Logger.getLogger(AddressHelper.class).finest("Address not available", ignored);
+        }
+
         int possiblePort = port;
         if (possiblePort == -1) {
             possiblePort = INITIAL_FIRST_PORT;
         }
         final Collection<Address> addresses = new LinkedList<Address>();
-        for (int i = 0; i < portTryCount; i++) {
-            try {
-                addresses.add(new Address(scopedAddress, possiblePort + i));
-            } catch (UnknownHostException ignored) {
-                Logger.getLogger(AddressHelper.class).finest("Address not available", ignored);
+
+        if (inetAddress == null) {
+            for (int i = 0; i < portTryCount; i++) {
+                try {
+                    addresses.add(new Address(scopedAddress, possiblePort + i));
+                } catch (UnknownHostException ignored) {
+                    Logger.getLogger(AddressHelper.class).finest("Address not available", ignored);
+                }
+            }
+        } else if (inetAddress instanceof Inet4Address) {
+            for (int i = 0; i < portTryCount; i++) {
+                addresses.add(new Address(scopedAddress, inetAddress, possiblePort + i));
+            }
+        } else if (inetAddress instanceof Inet6Address) {
+            final Collection<Inet6Address> possibleInetAddresses = getPossibleInetAddressesFor((Inet6Address) inetAddress);
+            for (Inet6Address inet6Address : possibleInetAddresses) {
+                for (int i = 0; i < portTryCount; i++) {
+                    addresses.add(new Address(scopedAddress, inet6Address, possiblePort + i));
+                }
             }
         }
 

--- a/hazelcast-client/src/main/java/com/hazelcast/client/util/AddressHelper.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/util/AddressHelper.java
@@ -17,18 +17,13 @@
 package com.hazelcast.client.util;
 
 import com.hazelcast.logging.Logger;
+import com.hazelcast.nio.Address;
 import com.hazelcast.util.AddressUtil;
 import com.hazelcast.util.AddressUtil.AddressHolder;
 
-import java.net.Inet4Address;
-import java.net.Inet6Address;
-import java.net.InetAddress;
-import java.net.InetSocketAddress;
 import java.net.UnknownHostException;
 import java.util.Collection;
 import java.util.LinkedList;
-
-import static com.hazelcast.util.AddressUtil.getPossibleInetAddressesFor;
 
 /**
  * This is a client side utility class for working with addresses and cluster connections
@@ -41,7 +36,7 @@ public final class AddressHelper {
     private AddressHelper() {
     }
 
-    public static Collection<InetSocketAddress> getSocketAddresses(String address) {
+    public static Collection<Address> getSocketAddresses(String address) {
         final AddressHolder addressHolder = AddressUtil.getAddressHolder(address, -1);
         final String scopedAddress = addressHolder.getScopeId() != null
                 ? addressHolder.getAddress() + '%' + addressHolder.getScopeId()
@@ -55,37 +50,21 @@ public final class AddressHelper {
         return getPossibleSocketAddresses(port, scopedAddress, maxPortTryCount);
     }
 
-    public static Collection<InetSocketAddress> getPossibleSocketAddresses(int port, String scopedAddress,
+    public static Collection<Address> getPossibleSocketAddresses(int port, String scopedAddress,
                                                                            int portTryCount) {
-        InetAddress inetAddress = null;
-        try {
-            inetAddress = InetAddress.getByName(scopedAddress);
-        } catch (UnknownHostException ignored) {
-            Logger.getLogger(AddressHelper.class).finest("Address not available", ignored);
-        }
-
         int possiblePort = port;
         if (possiblePort == -1) {
             possiblePort = INITIAL_FIRST_PORT;
         }
-        final Collection<InetSocketAddress> socketAddresses = new LinkedList<InetSocketAddress>();
-        if (inetAddress == null) {
-            for (int i = 0; i < portTryCount; i++) {
-                socketAddresses.add(new InetSocketAddress(scopedAddress, possiblePort + i));
-            }
-        } else if (inetAddress instanceof Inet4Address) {
-            for (int i = 0; i < portTryCount; i++) {
-                socketAddresses.add(new InetSocketAddress(inetAddress, possiblePort + i));
-            }
-        } else if (inetAddress instanceof Inet6Address) {
-            final Collection<Inet6Address> addresses = getPossibleInetAddressesFor((Inet6Address) inetAddress);
-            for (Inet6Address inet6Address : addresses) {
-                for (int i = 0; i < portTryCount; i++) {
-                    socketAddresses.add(new InetSocketAddress(inet6Address, possiblePort + i));
-                }
+        final Collection<Address> addresses = new LinkedList<Address>();
+        for (int i = 0; i < portTryCount; i++) {
+            try {
+                addresses.add(new Address(scopedAddress, possiblePort + i));
+            } catch (UnknownHostException ignored) {
+                Logger.getLogger(AddressHelper.class).finest("Address not available", ignored);
             }
         }
 
-        return socketAddresses;
+        return addresses;
     }
 }

--- a/hazelcast-client/src/main/java/com/hazelcast/client/util/ClientStateListener.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/util/ClientStateListener.java
@@ -54,7 +54,7 @@ public class ClientStateListener
      *
      * @param clientConfig The client configuration to which this listener will be registered
      */
-    ClientStateListener(ClientConfig clientConfig) {
+    public ClientStateListener(ClientConfig clientConfig) {
         clientConfig.addListenerConfig(new ListenerConfig(this));
     }
 

--- a/hazelcast-client/src/test/java/com/hazelcast/client/ClientConnectionTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/ClientConnectionTest.java
@@ -106,12 +106,12 @@ public class ClientConnectionTest extends HazelcastTestSupport {
         config.setProperty(ClientProperty.SHUFFLE_MEMBER_LIST.getName(), "false");
         config.getNetworkConfig().setSmartRouting(false);
 
-        InetSocketAddress socketAddress1 = server1.getCluster().getLocalMember().getSocketAddress();
-        InetSocketAddress socketAddress2 = server2.getCluster().getLocalMember().getSocketAddress();
+        Address address1 = server1.getCluster().getLocalMember().getAddress();
+        Address address2 = server2.getCluster().getLocalMember().getAddress();
 
         config.getNetworkConfig().
-                addAddress(socketAddress1.getHostName() + ":" + socketAddress1.getPort()).
-                addAddress(socketAddress2.getHostName() + ":" + socketAddress2.getPort());
+                addAddress(address1.getHost() + ":" + address1.getPort()).
+                addAddress(address2.getHost() + ":" + address2.getPort());
 
         hazelcastFactory.newHazelcastClient(config);
 
@@ -133,7 +133,7 @@ public class ClientConnectionTest extends HazelcastTestSupport {
 
         connectionManager.addConnectionListener(listener);
 
-        final Address serverAddress = new Address(server.getCluster().getLocalMember().getSocketAddress());
+        final Address serverAddress = server.getCluster().getLocalMember().getAddress();
         final Connection connectionToServer = connectionManager.getActiveConnection(serverAddress);
 
         final CountDownLatch isConnected = new CountDownLatch(1);

--- a/hazelcast-client/src/test/java/com/hazelcast/client/ClientRegressionWithRealNetworkTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/ClientRegressionWithRealNetworkTest.java
@@ -17,12 +17,20 @@
 package com.hazelcast.client;
 
 import com.hazelcast.client.config.ClientConfig;
+import com.hazelcast.client.config.ClientNetworkConfig;
+import com.hazelcast.client.connection.ClientConnectionManager;
+import com.hazelcast.client.impl.HazelcastClientInstanceImpl;
+import com.hazelcast.client.test.ClientTestSupport;
+import com.hazelcast.client.util.ClientStateListener;
 import com.hazelcast.config.Config;
+import com.hazelcast.core.EntryEvent;
 import com.hazelcast.core.Hazelcast;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.core.IMap;
+import com.hazelcast.map.listener.EntryAddedListener;
+import com.hazelcast.spi.properties.GroupProperty;
+import com.hazelcast.test.AssertTask;
 import com.hazelcast.test.HazelcastSerialClassRunner;
-import com.hazelcast.test.HazelcastTestSupport;
 import com.hazelcast.test.annotation.QuickTest;
 import org.junit.After;
 import org.junit.Test;
@@ -32,13 +40,17 @@ import org.junit.runner.RunWith;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
 
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotEquals;
 import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
 
 @RunWith(HazelcastSerialClassRunner.class)
 @Category(QuickTest.class)
-public class ClientRegressionWithRealNetworkTest extends HazelcastTestSupport {
+public class ClientRegressionWithRealNetworkTest extends ClientTestSupport {
 
     @After
     public void cleanUp() {
@@ -90,5 +102,85 @@ public class ClientRegressionWithRealNetworkTest extends HazelcastTestSupport {
         });
 
         assertOpenEventually(clientLatch);
+    }
+
+    @Test
+    public void testConnectWithDNSHostnames() throws InterruptedException {
+        Config config = new Config();
+        config.getNetworkConfig().setPublicAddress("localhost");
+        HazelcastInstance hazelcastInstance = Hazelcast.newHazelcastInstance(config);
+
+        ClientConfig clientConfig = new ClientConfig();
+        clientConfig.getNetworkConfig().addAddress("localhost").setConnectionAttemptLimit(Integer.MAX_VALUE);
+        ClientStateListener clientStateListener = new ClientStateListener(clientConfig);
+        HazelcastInstance client = HazelcastClient.newHazelcastClient(clientConfig);
+        HazelcastClientInstanceImpl clientInstanceImpl = getHazelcastClientInstanceImpl(client);
+        final ClientConnectionManager connectionManager = clientInstanceImpl.getConnectionManager();
+
+        Hazelcast.newHazelcastInstance(config);
+
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run()
+                    throws Exception {
+                assertEquals(2, connectionManager.getActiveConnections().size());
+            }
+        });
+
+        hazelcastInstance.shutdown();
+
+        assertTrue(clientStateListener.awaitConnected(ASSERT_TRUE_EVENTUALLY_TIMEOUT, TimeUnit.SECONDS));
+
+        assertEquals(1, connectionManager.getActiveConnections().size());
+    }
+
+    @Test
+    public void testListenersWhenDNSHostnamesAreUsed() {
+        Config config = new Config();
+        int heartBeatSeconds = 5;
+        config.getNetworkConfig().setPublicAddress("localhost");
+        config.setProperty(GroupProperty.CLIENT_HEARTBEAT_TIMEOUT_SECONDS.getName(), Integer.toString(heartBeatSeconds));
+        HazelcastInstance hazelcastInstance = Hazelcast.newHazelcastInstance(config);
+
+        ClientConfig clientConfig = new ClientConfig();
+        ClientNetworkConfig networkConfig = clientConfig.getNetworkConfig();
+        networkConfig.addAddress("localhost");
+        networkConfig.setConnectionAttemptLimit(Integer.MAX_VALUE);
+        final HazelcastInstance client = HazelcastClient.newHazelcastClient(clientConfig);
+        final IMap<Integer, Integer> map = client.getMap("test");
+
+        final AtomicInteger eventCount = new AtomicInteger(0);
+
+        map.addEntryListener(new EntryAddedListener() {
+            @Override
+            public void entryAdded(EntryEvent event) {
+                eventCount.incrementAndGet();
+            }
+        }, false);
+
+        Hazelcast.newHazelcastInstance(config);
+
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() throws Exception {
+                HazelcastClientInstanceImpl clientInstanceImpl = getHazelcastClientInstanceImpl(client);
+                int size = clientInstanceImpl.getConnectionManager().getActiveConnections().size();
+                assertEquals(2, size);
+
+            }
+        });
+
+        hazelcastInstance.shutdown();
+
+        sleepAtLeastSeconds(2 * heartBeatSeconds);
+
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run()
+                    throws Exception {
+                map.put(1, 2);
+                assertNotEquals(0, eventCount.get());
+            }
+        });
     }
 }

--- a/hazelcast-client/src/test/java/com/hazelcast/client/ClientServiceTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/ClientServiceTest.java
@@ -362,8 +362,7 @@ public class ClientServiceTest extends ClientTestSupport {
             map.put(randomString(), randomString());
         }
         HazelcastClientInstanceImpl clientInstanceImpl = ClientTestUtil.getHazelcastClientInstanceImpl(client);
-        InetSocketAddress socketAddress = hazelcastInstance.getCluster().getLocalMember().getSocketAddress();
-        Address address = new Address(socketAddress.getAddress().getHostAddress(), socketAddress.getPort());
+        Address address = hazelcastInstance.getCluster().getLocalMember().getAddress();
         ClientConnectionManager connectionManager = clientInstanceImpl.getConnectionManager();
         final ClientConnection connection = (ClientConnection) connectionManager.getActiveConnection(address);
         assertTrueEventually(new AssertTask() {

--- a/hazelcast-client/src/test/java/com/hazelcast/client/replicatedmap/ClientReplicatedMapLiteMemberTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/replicatedmap/ClientReplicatedMapLiteMemberTest.java
@@ -23,6 +23,7 @@ import com.hazelcast.client.test.TestHazelcastFactory;
 import com.hazelcast.config.Config;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.core.ReplicatedMap;
+import com.hazelcast.nio.Address;
 import com.hazelcast.replicatedmap.ReplicatedMapCantBeCreatedOnLiteMemberException;
 import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.annotation.ParallelTest;
@@ -124,10 +125,10 @@ public class ClientReplicatedMapLiteMemberTest {
     }
 
     private void configureDummyClientConnection(HazelcastInstance instance) throws UnknownHostException {
-        InetSocketAddress socketAddress = getAddress(instance).getInetSocketAddress();
+        Address memberAddress = getAddress(instance);
         dummyClientConfig.setProperty(ClientProperty.SHUFFLE_MEMBER_LIST.getName(), "false");
         ClientNetworkConfig networkConfig = dummyClientConfig.getNetworkConfig();
-        networkConfig.addAddress(socketAddress.getHostName() + ":" + socketAddress.getPort());
+        networkConfig.addAddress(memberAddress.getHost() + ":" + memberAddress.getPort());
     }
 
     private List<HazelcastInstance> createNodes(int numberOfLiteNodes, int numberOfDataNodes) {

--- a/hazelcast-client/src/test/java/com/hazelcast/client/spi/impl/ClientConnectionManagerTranslateTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/spi/impl/ClientConnectionManagerTranslateTest.java
@@ -33,7 +33,7 @@ import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
-import java.net.InetSocketAddress;
+import java.net.UnknownHostException;
 import java.util.Collection;
 import java.util.Collections;
 
@@ -97,8 +97,12 @@ public class ClientConnectionManagerTranslateTest extends ClientTestSupport {
 
     private class TestAddressProvider implements AddressProvider {
         @Override
-        public Collection<InetSocketAddress> loadAddresses() {
-            return Collections.singletonList(new InetSocketAddress("127.0.0.1", 5701));
+        public Collection<Address> loadAddresses() {
+            try {
+                return Collections.singletonList(new Address("127.0.0.1", 5701));
+            } catch (UnknownHostException e) {
+                return null;
+            }
         }
     }
 

--- a/hazelcast-client/src/test/java/com/hazelcast/client/test/TestHazelcastFactory.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/test/TestHazelcastFactory.java
@@ -33,7 +33,6 @@ import com.hazelcast.spi.properties.HazelcastProperties;
 import com.hazelcast.test.TestEnvironment;
 import com.hazelcast.test.TestHazelcastInstanceFactory;
 
-import java.net.InetSocketAddress;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.List;
@@ -104,14 +103,14 @@ public class TestHazelcastFactory extends TestHazelcastInstanceFactory {
 
         return new AddressProvider() {
             @Override
-            public Collection<InetSocketAddress> loadAddresses() {
-                Collection<InetSocketAddress> inetAddresses = new ArrayList<InetSocketAddress>();
+            public Collection<Address> loadAddresses() {
+                Collection<Address> possibleAddresses = new ArrayList<Address>();
                 for (Address address : getKnownAddresses()) {
-                    Collection<InetSocketAddress> addresses = AddressHelper.getPossibleSocketAddresses(address.getPort(),
+                    Collection<Address> addresses = AddressHelper.getPossibleSocketAddresses(address.getPort(),
                             address.getHost(), 1);
-                    inetAddresses.addAll(addresses);
+                    possibleAddresses.addAll(addresses);
                 }
-                return inetAddresses;
+                return possibleAddresses;
             }
         };
     }

--- a/hazelcast-client/src/test/java/com/hazelcast/client/test/bounce/UnisocketClientDriverFactory.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/test/bounce/UnisocketClientDriverFactory.java
@@ -19,6 +19,7 @@ package com.hazelcast.client.test.bounce;
 import com.hazelcast.client.config.ClientConfig;
 import com.hazelcast.client.spi.properties.ClientProperty;
 import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.nio.Address;
 
 import java.net.InetSocketAddress;
 
@@ -46,10 +47,9 @@ public class UnisocketClientDriverFactory extends AbstractClientDriverFactory {
         config.setProperty(ClientProperty.SHUFFLE_MEMBER_LIST.getName(), "false");
         config.getNetworkConfig().setSmartRouting(false);
 
-        InetSocketAddress socketAddress = member.getCluster().getLocalMember().getSocketAddress();
+        Address address = member.getCluster().getLocalMember().getAddress();
 
-        config.getNetworkConfig().
-                addAddress(socketAddress.getHostName() + ":" + socketAddress.getPort());
+        config.getNetworkConfig().addAddress(address.getHost() + ":" + address.getPort());
 
         return config;
     }

